### PR TITLE
Async get and get bulk

### PR DESCRIPTION
**Pull request by sriram-rangarajan**
@tharanga Please review

### CODE REVIEW

File: evcache-core/src/test/java/com/netflix/evcache/test/Base.java [hunk 1]
```diff
@@ -239,6 +248,13 @@ public Map<String, String> getBulk(String keys[], EVCache gCache) throws Excepti
         return value;
     }
 
+    public Map<String, String> getCompletableBulk(String keys[], EVCache gCache) throws Exception {
+        ExecutorService executorService = Executors.newFixedThreadPool(10);
+        final CompletableFuture<Map<String, String>> value = gCache.<String>getBulk(executorService, keys);
+        if(log.isDebugEnabled()) log.debug("getBulk : keys : " + Arrays.toString(keys) + "; values = " + value);
+        return value.get();
```
Comment (tharanga) [inline evcache-core/src/test/java/com/netflix/evcache/test/Base.java:255]: Can we add a test case that would use this in an async manner? You have to wait at a different level to ensure results are returned as expected, but not using the `get` method.

File: evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java [hunk 2]
```diff
@@ -512,12 +533,167 @@ <T> T doGet(EVCacheKey evcKey , Transcoder<T> tc) throws EVCacheException {
...
+                .thenCompose(data -> handleRetry(data, evcKey, tc, client, hasZF, throwExc, event, executorService))
+                .thenApply(data -> handleData(data, event, evcKey, client, cacheOperation))
+                .exceptionally(ex -> handleException(ex, throwExc, event, evcKey))
+                .handle((data, ex) -> handleFinally(data, status, tries, client, cacheOperation, start));
+    }
+
+    private EVCacheClient buildEvCacheClient(boolean throwExc) throws EVCacheException {
+        EVCacheClient client = _pool.getEVCacheClientForRead();
+        if (client == null) {
+            incrementFastFail(EVCacheMetricsFactory.NULL_CLIENT, Call.COMPLETABLE_FUTURE_GET);
```
Comment (sumanth-pasupuleti) [inline evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java]: method "buildEvCacheClient" seems generic (beyond just Completable-Future-Get). Semantically, it seems odd this method increments null client for CompletableFutureGet.
My recommendation - either take the Call enum as argument in the method, or not create this method and build new evcache inline within "doAsyncGet" method

File: evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java [hunk 3]
```diff
@@ -512,12 +533,167 @@ <T> T doGet(EVCacheKey evcKey , Transcoder<T> tc) throws EVCacheException {
...
+        if (null == key) throw new IllegalArgumentException("Key cannot be null");
+        final EVCacheKey evcKey = getEVCacheKey(key);
+        return doAsyncGet(evcKey, tc, executorService);
+    }
+
+    <T> CompletableFuture<T> doAsyncGet(EVCacheKey evcKey, Transcoder<T> tc, ExecutorService executorService) throws EVCacheException {
+        final boolean throwExc = doThrowException();
+        EVCacheClient client = buildEvCacheClient(throwExc);
+        if (client == null) {
+            return null;
```
[the remark this answers is not in the document]
Comment (sumanth-pasupuleti) [reply evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java]: I see that you do this inside method "buildEvCacheClient" - I put  a related comment in that method.

File: evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java [hunk 4]
```diff
@@ -512,12 +533,167 @@ <T> T doGet(EVCacheKey evcKey , Transcoder<T> tc) throws EVCacheException {
...
+        return client;
+    }
+
+    private EVCacheEvent buildAndStartEvent(EVCacheClient client, EVCacheKey evcKey, boolean throwExc) throws EVCacheException {
+        EVCacheEvent event = createEVCacheEvent(Collections.singletonList(client), Call.COMPLETABLE_FUTURE_GET);
+        if (event != null) {
+            event.setEVCacheKeys(Arrays.asList(evcKey));
+            try {
+                if (shouldThrottle(event)) {
+                    incrementFastFail(EVCacheMetricsFactory.THROTTLED, Call.COMPLETABLE_FUTURE_GET);
```
Comment (sumanth-pasupuleti) [inline evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java:620]: similar comment as buildEvCacheClient method. This method name seems generic, but it increments metric against COMPLETABLE_FUTURE_GET

File: evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java [hunk 5]
```diff
@@ -512,12 +533,167 @@ <T> T doGet(EVCacheKey evcKey , Transcoder<T> tc) throws EVCacheException {
...
+            }
+            startEvent(event);
+            return event;
+        }
+        return null;
+    }
+
+    private <T> T handleFinally(T data, StringBuilder status, AtomicInteger tries, EVCacheClient client, StringBuilder cacheOperation, Long start) {
+        final long duration = EVCacheMetricsFactory.getInstance().getRegistry().clock().wallTime() - start;
+        getTimer(Call.COMPLETABLE_FUTURE_GET.name(),
```
Comment (sumanth-pasupuleti) [inline evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java]: similar comment as buildEvCacheClient

File: evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java [hunk 6]
```diff
@@ -512,12 +533,167 @@ <T> T doGet(EVCacheKey evcKey , Transcoder<T> tc) throws EVCacheException {
...
+        if (data != null) {
+            if (event != null) event.setAttribute("status", "GHIT");
+        } else {
+            cacheOperation.replace(0, cacheOperation.length(), EVCacheMetricsFactory.NO);
+            if (event != null) event.setAttribute("status", "GMISS");
+            if (log.isInfoEnabled() && shouldLog())
+                log.info("GET : APP " + _appName + " ; cache miss for key : " + evcKey);
+        }
+        if (log.isDebugEnabled() && shouldLog())
+            log.debug("COMPLETABLE FUTURE GET : APP " + _appName + ", key [" + evcKey + (log.isTraceEnabled() ? "], Value [" + data : "") + "], ServerGroup : " + client.getServerGroup());
```
Comment (sumanth-pasupuleti) [inline evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java:704]: similar comment as buildEvCacheClient

File: evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java [hunk 7]
```diff
@@ -1518,80 +1829,199 @@ public T get(long timeout, TimeUnit unit) throws InterruptedException, Execution
...
+                hasHashedKey = true;
+            }
+            keyMap.put(key, evcKey);
+        }
+        return new KeyMapDto(keyMap, hasHashedKey);
+    }
+
+    private <T>  Map<EVCacheKey, T>  buildNonHashedKeyValueResult(Map<String, T> objMap,
+                                                                  Map<String, EVCacheKey> keyMap) {
+        final Map<EVCacheKey, T> retMap = new HashMap<EVCacheKey, T>((int) (objMap.size() / 0.75) + 1);
```
Comment (sumanth-pasupuleti) [inline evcache-core/src/main/java/com/netflix/evcache/EVCacheImpl.java:1924]: curious on 0.75. I see this is from existing code, but maybe good to add a comment about why 0.75

File: evcache-core/src/test/java/com/netflix/evcache/test/SimpleEVCacheTest.java [hunk 8]
```diff
@@ -32,8 +34,8 @@
 public class SimpleEVCacheTest extends Base {
     private static final Logger log = LogManager.getLogger(SimpleEVCacheTest.class);
 
-    private static final String APP_NAME = "EVCACHE_TEST";
-    private static final String ALIAS_APP_NAME = "EVCACHE";
+    private static final String APP_NAME = "EVCACHE_TEST_SRIRAM";
```
Comment (sumanth-pasupuleti) [inline evcache-core/src/test/java/com/netflix/evcache/test/SimpleEVCacheTest.java:37]: if this warrants needing a real cluster, it maybe useful to have (and refer) a generic integration test cluster (like evcache_cde_integration_test)

File: evcache-core/src/test/java/com/netflix/evcache/test/SimpleEVCacheTest.java [hunk 9]
```diff
@@ -172,6 +174,16 @@ public void testGet() throws Exception {
         }
     }
 
+
+    @Test(dependsOnMethods = { "testInsert" })
+    public void testCompletableFutureGet() throws Exception {
+        for (int i = 0; i < 1000; i++) {
+            final String val = completableFutureGet(i, evCache);
+            //assertNotNull(val);
```
Comment (sumanth-pasupuleti) [inline evcache-core/src/test/java/com/netflix/evcache/test/SimpleEVCacheTest.java:182]: not sure if this is complete?

File: evcache-core/src/test/java/com/netflix/evcache/test/SimpleEVCacheTest.java [hunk 10]
```diff
@@ -202,6 +214,20 @@ public void testBulk() throws Exception {
...
+    public void testCompletableFutureBulk() throws Exception {
+        final String[] keys = new String[12];
+        for (int i = 0; i < keys.length; i++) {
+            keys[i] = "key_" + i;
+        }
+        Map<String, String> vals = getAsyncBulk(keys, evCache);
+        assertTrue(!vals.isEmpty());
+        for (int i = 0; i < vals.size(); i++) {
+            String key = "key_" + i;
+            String val = vals.get(key);
```
Comment (sumanth-pasupuleti) [inline evcache-core/src/test/java/com/netflix/evcache/test/SimpleEVCacheTest.java:227]: if we aren't validating "val", maybe ignore defining the val variable? I am assuming the intention of vals.get(key) is that the dictionary has a key for every "key" we pass into getAsyncBulk() method - is that correct?

**Review by sumanth-pasupuleti — APPROVED** [review]

**Pull request merged**